db/keyval/filesystem/reader/reader.go

```diff
@@ -0,0 +1,189 @@
+//  Copyright (c) 2018 Cisco and/or its affiliates.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at:
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package reader
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+// API of file system reader
+type API interface {
+	// PathExists returns true if provided file exists, false otherwise
+	PathExists(path string) bool
+	// IsDirectory returns true if path is directory, false if it is a file
+	IsDirectory(path string) (bool, error)
+	// ProcessFile returns File type object. Provided path must be a file
+	ProcessFile(path string) (File, error)
+	// ProcessFilesInDir returns list of File type objects. Provided path must be a directory
+	ProcessFilesInDir(path string) ([]File, error)
+	// IsValid validates interface
+	IsValid(ev fsnotify.Event) (bool, error)
+}
+
+// Reader implements file system API
+type Reader struct{}
```

Comment:
It seems to me that this reader API implementation should actually be called `JsonReader`, because it's tightly coupled with json encoding/decoding at the moment.

Reply:
API changed, it is not coupled with the reader implementation anymore.